src/disjoint_set.jl

```diff
@@ -57,22 +57,28 @@ function union!(s::IntDisjointSets, x::Integer, y::Integer)
     parents = s.parents
     xroot = find_root_impl!(parents, x)
     yroot = find_root_impl!(parents, y)
-
     if xroot != yroot
-        rks = s.ranks
-        @inbounds xrank = rks[xroot]
-        @inbounds yrank = rks[yroot]
-
-        if xrank < yrank
-            @inbounds parents[xroot] = yroot
-        else
-            @inbounds parents[yroot] = xroot
-            if xrank == yrank
-                rks[xroot] += 1
-            end
-        end
-        @inbounds s.ngroups -= 1
+        _union!(s, xroot, yroot)
+    end
+end
+
+# form a new set that is the union of the two sets whose root elements are
+# x and y and return the root of the new set
+# assume x ≠ y (unsafe)
+function _union!(s::IntDisjointSets, x::Integer, y::Integer)
+    parents = s.parents
+    rks = s.ranks
+    @inbounds xrank = rks[x]
+    @inbounds yrank = rks[y]
+
+    if xrank < yrank
+        x, y = y, x
+    elseif xrank == yrank
+        rks[x] += 1
     end
+    @inbounds parents[y] = x
+    @inbounds s.ngroups -= 1
+    x
 end
```

Comment:
In Julia, it's probably not a good idea to call a function which isn't part of the official interface, especially one with an underscore as the first character.  How about calling it `root_union!` (or some other name), documenting and exporting it?

Of course, then it should also be implemented for `DisjointSets`.